README.md


# HelpMe!

Add flexiable user guide for pages of your application!

Example:

![](https://github.com/mahdices/flutter_help_me/blob/main/example/sample/sample.gif)

## Usage

```dart
HelpMe.showHelp(
            color: Colors.amber.withOpacity(0.2),
            gradient: LinearGradient(colors: [
              Colors.purple.withOpacity(0.2),
              Colors.blue.withOpacity(0.2),
            ]),items: [
            HelpMeItem(
                    key: keyButton,
                    shape: const HelpMeRectShape(
                        borderRadius: BorderRadius.all(Radius.circular(24))),
                    guideWidget: Padding(
                      padding: const EdgeInsets.all(16.0),
                      child: Text(
                        "Lorem Guide",
                        style: TextStyle(color: Colors.white),
                      ),
                    )),
                    ...
          ], pageContext: context, nextWidget: Text("Next"), skipWidget: Text("Skip"));
```

## Notable References 

- [Documentation for Key in flutter](https://api.flutter.dev/flutter/foundation/Key-class.html)
- [Documentation for CustomPainter in flutter](https://api.flutter.dev/flutter/rendering/CustomPainter-class.html)
- [Documentation for CustomClipper in flutter](https://api.flutter.dev/flutter/rendering/CustomClipper-class.html)

## End

If you have some suggestion or advice, please open an issue or develop the fix and open a pull request. This will greatly help the improvement of the usability of this project. Thanks.

CHANGELOG.md
# Changelog

## 0.0.1

  * Initial version release
  * Release ready, very basic functionality

example/lib/custom_shape_test.dart
import 'package:flutter/material.dart';

class RPSCustomPainter2 extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    Paint paint0 = Paint()
      ..color = const Color.fromARGB(255, 33, 150, 243)
      ..style = PaintingStyle.stroke
      ..strokeWidth = 1;

    canvas.drawPath(getPath(size), paint0);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }
}

Path getPath(Size size) {
  Path path0 = Path();
  path0.moveTo(size.width * 0.4051917, size.height * 0.3986763);
  path0.lineTo(size.width * 0.5550000, size.height * 0.2158273);
  path0.lineTo(size.width * 0.6783333, size.height * 0.4474820);
  path0.lineTo(size.width * 0.5308333, size.height * 0.6043165);
  path0.lineTo(size.width * 0.4051917, size.height * 0.3986763);
  path0.close();
  return path0;
}


lib/help_me.dart
export 'src/guide.dart';

import 'package:flutter/material.dart';
import 'package:help_me/src/guide.dart';
import 'package:help_me/src/help.dart';

class HelpMe {
  static showHelp(
      {required List<HelpMeItem> items,
      required BuildContext pageContext,
      required Widget nextWidget,
      Color? color,
      Gradient? gradient,
      required Widget skipWidget}) {
    showModalBottomSheet(
        elevation: 0,
        context: pageContext,
        backgroundColor: Colors.transparent,
        isScrollControlled: true,
        barrierColor: Colors.transparent,
        builder: (_) {
          return Help(
            nextButton: nextWidget,
            skipButton: skipWidget,
            guides: items,
            gradient: gradient,
            color: color,
            pageContext: pageContext,
          );
        });
  }
}


lib/src/help.dart
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

import 'guide.dart';
import 'help_me_clipper.dart';

class Help extends StatefulWidget {
  List<HelpMeItem> guides;
  Size? size;
  BuildContext pageContext;
  Widget? nextButton;
  Widget? skipButton;
  final Color? color;
  final Gradient? gradient;

  Help(
      {Key? key,
      required this.guides,
      required this.pageContext,
      this.nextButton,
      this.skipButton,
      this.color,
      this.gradient})
      : super(key: key) {
    size = Size(MediaQuery.of(pageContext).size.width,
        MediaQuery.of(pageContext).size.height);
  }

  @override
  State<Help> createState() => _HelpState();
}

class _HelpState extends State<Help> {
  int selected = 0;
  bool isShowing = true;
  RenderBox? currentBox;
  Offset get currentPosition =>
      currentBox?.localToGlobal(Offset.zero) ?? Offset(0, 0);
  RenderBox? nextButtonBox;
  // RenderBox? guidBox;
  GlobalKey? skipButtonKey;
  var nextKey = GlobalKey();
  var oldSize;

  List<Widget> getItemsOnPage(bool reverse) {
    List<Widget> list = [];
    if (reverse) {
      list.add(_buttons());
      list.add(_guid());
    } else {
      list.add(_guid());
      list.add(_buttons());
    }

    return list;
  }

  _nextGuide() async {
    // guidBox = widget.guides[selected].guidKey.currentContext!.findRenderObject()
    //     as RenderBox;

    // await Future.delayed(Duration(milliseconds: 1000));
    if (widget.guides.length - 1 == selected) {
      _skip();
      return;
    }
    setState(() {
      selected++;
      currentBox = widget.guides[selected].key.currentContext!
          .findRenderObject() as RenderBox;
    });
  }

  _skip() {
    Navigator.of(context).pop();
  }

  _buttons() {
    if (widget.nextButton == null) {
      return Container();
    }

    var align = Alignment.bottomCenter;
    if (nextButtonBox != null) {
      var nextButtonPos = nextButtonBox!.localToGlobal(Offset.zero).dy;
      // var guidPos = guidBox!.localToGlobal(Offset.zero).dy;
      if (currentPosition.dy + (currentBox?.size.height ?? 0) >
          (nextButtonPos + ((widget.size!.height / 2) - nextButtonPos))) {
        align = Alignment.topCenter;
      }
    }

    return AnimatedAlign(
      duration: Duration(milliseconds: 200),
      alignment: align,
      child: Padding(
        padding: MediaQuery.of(widget.pageContext).padding,
        child: Row(
          mainAxisSize: MainAxisSize.min,
          children: [
            widget.skipButton != null
                ? GestureDetector(
                    child: widget.skipButton,
                    onTap: () {
                      _skip();
                    },
                  )
                : Container(),
            GestureDetector(
              child: widget.nextButton,
              onTap: () {
                _nextGuide();
              },
            ),
          ],
        ),
      ),
    );
  }

  _guid() {
    var buttonIsBottom = true;
    double buttonHeight = 0;
    double guideHeight = 0;

    double guideTopPadding = 0;
    double guideBottomPadding = 0;

    if (nextButtonBox != null) {
      buttonHeight = nextButtonBox!.size.height;
      guideHeight = currentBox!.size.height;
      if (currentPosition.dy + (currentBox?.size.height ?? 0) >
          (nextButtonBox!.localToGlobal(Offset.zero).dy +
              ((widget.size!.height / 2) -
                  nextButtonBox!.localToGlobal(Offset.zero).dy))) {
        buttonIsBottom = false;
      }
    }
    var align = Alignment.bottomCenter;

    if (buttonIsBottom) {
      if (widget.size!.height -
              currentPosition.dy +
              (currentBox?.size.height ?? 0) -
              buttonHeight >
          currentPosition.dy) {
        guideBottomPadding = buttonHeight;
        align = Alignment.bottomCenter;
      } else {
        align = Alignment.topCenter;
      }
    } else {
      // change if button is top and text is bellow state 2
      guideTopPadding = buttonHeight;
      align = Alignment.topCenter;
    }

    return AnimatedAlign(
      duration: Duration(milliseconds: 200),
      alignment: align,
      child: Padding(
        padding: MediaQuery.of(widget.pageContext).padding,
        child: AnimatedSwitcher(
          duration: Duration(milliseconds: 200),
          child: isShowing
              ? Padding(
                  padding: EdgeInsets.only(
                      top: guideTopPadding, bottom: guideBottomPadding),
                  child: widget.guides[selected].guideWidget,
                )
              : Container(),
        ),
      ),
    );
  }

  @override
  void initState() {
    currentBox = widget.guides[selected].key.currentContext!.findRenderObject()
        as RenderBox;

    widget.nextButton = Container(
      child: widget.nextButton,
      key: nextKey,
    );
    Future.delayed(Duration(milliseconds: 200), () => postFrameCallback());
    // WidgetsBinding.instance?.addPostFrameCallback((_) => postFrameCallback());
    super.initState();
  }

  void postFrameCallback() {
    var context = nextKey.currentContext;
    if (context == null) return;
    var newSize = context.findRenderObject() as RenderBox;

    setState(() {
      nextButtonBox = newSize;
      // guidBox = newSizeGuid;
      var nextButtonPos = nextButtonBox!.localToGlobal(Offset.zero).dy;
      // var guidPos = guidBox!.localToGlobal(Offset.zero).dy;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      width: double.infinity,
      height: double.infinity,
      child: Stack(
        children: [
          AnimatedSwitcher(
            duration: Duration(milliseconds: 200),
            child: Builder(builder: (context) {
              if (widget.size != null) {
                if (currentPosition.dy < 0 ||
                    currentPosition.dy > widget.size!.height) {
                  isShowing = false;
                  _moveScroll(currentBox!);
                  // widget.scrollController!.jumpTo(
                  //   position.dy,
                  // );

                }
              }

              return isShowing
                  ? Container(
                      child: AnimatedSwitcher(
                        duration: Duration(milliseconds: 300),
                        child: CustomPaint(
                          key: ValueKey(widget.guides[selected].key.hashCode),
                          painter: HelpMePainter(
                              widget.guides[selected].key, currentBox!,
                              shape: widget.guides[selected].shape),
                          child: ClipPath(
                            clipper: HelpMeClipper(
                                widget.guides[selected].key, currentBox!,
                                shape: widget.guides[selected].shape),
                            key: ValueKey(
                                widget.guides[selected].key.hashCode + 1),
                            child: Container(
                              decoration: BoxDecoration(
                                  color: widget.gradient == null
                                      ? (widget.color ??
                                          Colors.black.withOpacity(0.01))
                                      : null,
                                  gradient: widget.gradient),
                            ),
                          ),
                        ),
                      ),
                    )
                  : Container();
            }),
          ),
          // if (currentPosition.dy < 0 ||
          //     currentPosition.dy > widget.size!.height)
          //   Container()
          // else
          _buttons(),
          _guid()
        ],
      ),
    );
  }

  _moveScroll(
    RenderBox box,
  ) async {
    // Offset position = box.localToGlobal(Offset.zero);
    // double scroll = 0;
    // if (position.dy > widget.size!.height) {
    //   scroll = position.dy -
    //       widget.scrollController!.offset +
    //       box.paintBounds.height;
    // } else {
    //   scroll = position.dy + widget.scrollController!.offset;
    // }
    // print(scroll);
    // widget.scrollController!.jumpTo(scroll);
    Scrollable.ensureVisible(widget.guides[selected].key.currentContext!,
        duration: Duration(milliseconds: 200), curve: Curves.easeIn);
    // widget.scrollController!.animateTo(
    //   scroll,
    //   duration: Duration(milliseconds: 200),
    //   curve: Curves.easeIn,
    // );
    await Future.delayed(Duration(milliseconds: 600));
    setState(() {
      isShowing = true;
    });
  }
}


lib/src/guide.dart
import 'package:flutter/material.dart';

class HelpMeItem {
  final GlobalKey key;
  final Widget guideWidget;
  final HelpMeShape shape;

  HelpMeItem(
      {required this.key,
      this.shape = const HelpMeRectShape(),
      required this.guideWidget});
}

abstract class HelpMeShape {
  final double width;
  final Color color;
  const HelpMeShape({this.width = 3, this.color = Colors.white});
}

class HelpMeCustomShape extends HelpMeShape {
  final double width;
  final Color color;
  final Path path;

  HelpMeCustomShape(
      {required this.path, this.width = 3, this.color = Colors.white})
      : super(color: color, width: width);
}

class HelpMeOvalShape extends HelpMeShape {
  final double width;
  final Color color;
  const HelpMeOvalShape({this.width = 3, this.color = Colors.white})
      : super(color: color, width: width);
}

class HelpMeRectShape extends HelpMeShape {
  final double width;
  final Color color;
  final BorderRadius borderRadius;
  const HelpMeRectShape(
      {this.borderRadius = const BorderRadius.all(Radius.circular(8)),
      this.width = 3,
      this.color = Colors.white})
      : super(color: color, width: width);
}


example/lib/new_svg.dart
import 'dart:ui';

import 'package:flutter/material.dart';

//Copy this CustomPainter code to the Bottom of the File
class RPSCustomPainter extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    Path path_0 = Path();
    path_0.moveTo(size.width * 0.8535523, size.height * 0.1530197);
    path_0.cubicTo(
        size.width * 0.7591158,
        size.height * 0.05858110,
        size.width * 0.6335546,
        size.height * 0.006572009,
        size.width * 0.4999990,
        size.height * 0.006572009);
    path_0.lineTo(size.width * 0.4999990, size.height * 0.08785845);
    path_0.cubicTo(
        size.width * 0.7308789,
        size.height * 0.08785845,
        size.width * 0.9187115,
        size.height * 0.2756931,
        size.width * 0.9187115,
        size.height * 0.5065730);
    path_0.lineTo(size.width * 0.9187115, size.height * 0.5326233);
    path_0.cubicTo(
        size.width * 0.9688206,
        size.height * 0.5814338,
        size.width * 0.9999980,
        size.height * 0.6495884,
        size.width * 0.9999980,
        size.height * 0.7249003);
    path_0.lineTo(size.width * 0.9999980, size.height * 0.5065730);
    path_0.cubicTo(
        size.width,
        size.height * 0.3730174,
        size.width * 0.9479909,
        size.height * 0.2474562,
        size.width * 0.8535523,
        size.height * 0.1530197);
    path_0.close();

    Paint paint_0_fill = Paint()..style = PaintingStyle.fill;
    paint_0_fill.color = Color(0xff005ECE).withOpacity(1.0);
    canvas.drawPath(path_0, paint_0_fill);

    Path path_1 = Path();
    path_1.moveTo(size.width * 0.1464456, size.height * 0.1530197);
    path_1.cubicTo(size.width * 0.05200910, size.height * 0.2474562, 0,
        size.height * 0.3730174, 0, size.height * 0.5065730);
    path_1.lineTo(0, size.height * 0.7249003);
    path_1.lineTo(size.width * 0.3498142, size.height * 0.7249003);
    path_1.lineTo(size.width * 0.3498142, size.height * 0.4563705);
    path_1.lineTo(size.width * 0.2685277, size.height * 0.4563705);
    path_1.cubicTo(
        size.width * 0.1957723,
        size.height * 0.4563705,
        size.width * 0.1296945,
        size.height * 0.4854690,
        size.width * 0.08128644,
        size.height * 0.5326213);
    path_1.lineTo(size.width * 0.08128644, size.height * 0.5065710);
    path_1.cubicTo(
        size.width * 0.08128644,
        size.height * 0.2756911,
        size.width * 0.2691211,
        size.height * 0.08785642,
        size.width * 0.4999990,
        size.height * 0.08785642);
    path_1.lineTo(size.width * 0.4999990, size.height * 0.006569976);
    path_1.cubicTo(
        size.width * 0.3664454,
        size.height * 0.006572009,
        size.width * 0.2408842,
        size.height * 0.05858110,
        size.width * 0.1464456,
        size.height * 0.1530197);
    path_1.close();

    Paint paint_1_fill = Paint()..style = PaintingStyle.fill;
    paint_1_fill.color = Color(0xff2488FF).withOpacity(1.0);
    canvas.drawPath(path_1, paint_1_fill);

    Path path_2 = Path();
    path_2.moveTo(0, size.height * 0.7249003);
    path_2.cubicTo(
        0,
        size.height * 0.8729676,
        size.width * 0.1204604,
        size.height * 0.9934280,
        size.width * 0.2685277,
        size.height * 0.9934280);
    path_2.lineTo(size.width * 0.3498142, size.height * 0.9934280);
    path_2.lineTo(size.width * 0.3498142, size.height * 0.7249003);
    path_2.lineTo(0, size.height * 0.7249003);
    path_2.close();

    Paint paint_2_fill = Paint()..style = PaintingStyle.fill;
    paint_2_fill.color = Color(0xff005ECE).withOpacity(1.0);
    canvas.drawPath(path_2, paint_2_fill);

    Path path_3 = Path();
    path_3.moveTo(size.width * 0.9187136, size.height * 0.5326233);
    path_3.cubicTo(
        size.width * 0.8703055,
        size.height * 0.4854711,
        size.width * 0.8042277,
        size.height * 0.4563726,
        size.width * 0.7314723,
        size.height * 0.4563726);
    path_3.lineTo(size.width * 0.6501858, size.height * 0.4563726);
    path_3.lineTo(size.width * 0.6501858, size.height * 0.7249003);
    path_3.lineTo(size.width, size.height * 0.7249003);
    path_3.cubicTo(
        size.width,
        size.height * 0.6495884,
        size.width * 0.9688226,
        size.height * 0.5814317,
        size.width * 0.9187136,
        size.height * 0.5326233);
    path_3.close();

    Paint paint_3_fill = Paint()..style = PaintingStyle.fill;
    paint_3_fill.color = Color(0xff00479B).withOpacity(1.0);
    canvas.drawPath(path_3, paint_3_fill);

    Path path_4 = Path();
    path_4.moveTo(size.width * 0.6501858, size.height * 0.7249003);
    path_4.lineTo(size.width * 0.6501858, size.height * 0.9934280);
    path_4.lineTo(size.width * 0.7314723, size.height * 0.9934280);
    path_4.cubicTo(size.width * 0.8795396, size.height * 0.9934280, size.width,
        size.height * 0.8729676, size.width, size.height * 0.7249003);
    path_4.lineTo(size.width * 0.6501858, size.height * 0.7249003);
    path_4.close();

    Paint paint_4_fill = Paint()..style = PaintingStyle.fill;
    paint_4_fill.color = Color(0xff003068).withOpacity(1.0);
    canvas.drawPath(path_4, paint_4_fill);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }
}

Path getPath2(Size size) {
  var path = Path();
  Path path_0 = Path();
  path_0.moveTo(size.width * 0.8535523, size.height * 0.1530197);
  path_0.cubicTo(
      size.width * 0.7591158,
      size.height * 0.05858110,
      size.width * 0.6335546,
      size.height * 0.006572009,
      size.width * 0.4999990,
      size.height * 0.006572009);
  path_0.lineTo(size.width * 0.4999990, size.height * 0.08785845);
  path_0.cubicTo(
      size.width * 0.7308789,
      size.height * 0.08785845,
      size.width * 0.9187115,
      size.height * 0.2756931,
      size.width * 0.9187115,
      size.height * 0.5065730);
  path_0.lineTo(size.width * 0.9187115, size.height * 0.5326233);
  path_0.cubicTo(
      size.width * 0.9688206,
      size.height * 0.5814338,
      size.width * 0.9999980,
      size.height * 0.6495884,
      size.width * 0.9999980,
      size.height * 0.7249003);
  path_0.lineTo(size.width * 0.9999980, size.height * 0.5065730);
  path_0.cubicTo(size.width, size.height * 0.3730174, size.width * 0.9479909,
      size.height * 0.2474562, size.width * 0.8535523, size.height * 0.1530197);
  path_0.close();
  path.addPath(path_0, Offset.zero);
  Path path_1 = Path();
  path_1.moveTo(size.width * 0.1464456, size.height * 0.1530197);
  path_1.cubicTo(size.width * 0.05200910, size.height * 0.2474562, 0,
      size.height * 0.3730174, 0, size.height * 0.5065730);
  path_1.lineTo(0, size.height * 0.7249003);
  path_1.lineTo(size.width * 0.3498142, size.height * 0.7249003);
  path_1.lineTo(size.width * 0.3498142, size.height * 0.4563705);
  path_1.lineTo(size.width * 0.2685277, size.height * 0.4563705);
  path_1.cubicTo(
      size.width * 0.1957723,
      size.height * 0.4563705,
      size.width * 0.1296945,
      size.height * 0.4854690,
      size.width * 0.08128644,
      size.height * 0.5326213);
  path_1.lineTo(size.width * 0.08128644, size.height * 0.5065710);
  path_1.cubicTo(
      size.width * 0.08128644,
      size.height * 0.2756911,
      size.width * 0.2691211,
      size.height * 0.08785642,
      size.width * 0.4999990,
      size.height * 0.08785642);
  path_1.lineTo(size.width * 0.4999990, size.height * 0.006569976);
  path_1.cubicTo(
      size.width * 0.3664454,
      size.height * 0.006572009,
      size.width * 0.2408842,
      size.height * 0.05858110,
      size.width * 0.1464456,
      size.height * 0.1530197);
  path_1.close();

  path.addPath(path_1, Offset.zero);

  Path path_2 = Path();
  path_2.moveTo(0, size.height * 0.7249003);
  path_2.cubicTo(0, size.height * 0.8729676, size.width * 0.1204604,
      size.height * 0.9934280, size.width * 0.2685277, size.height * 0.9934280);
  path_2.lineTo(size.width * 0.3498142, size.height * 0.9934280);
  path_2.lineTo(size.width * 0.3498142, size.height * 0.7249003);
  path_2.lineTo(0, size.height * 0.7249003);
  path_2.close();
  path.addPath(path_2, Offset.zero);

  Path path_3 = Path();
  path_3.moveTo(size.width * 0.9187136, size.height * 0.5326233);
  path_3.cubicTo(
      size.width * 0.8703055,
      size.height * 0.4854711,
      size.width * 0.8042277,
      size.height * 0.4563726,
      size.width * 0.7314723,
      size.height * 0.4563726);
  path_3.lineTo(size.width * 0.6501858, size.height * 0.4563726);
  path_3.lineTo(size.width * 0.6501858, size.height * 0.7249003);
  path_3.lineTo(size.width, size.height * 0.7249003);
  path_3.cubicTo(size.width, size.height * 0.6495884, size.width * 0.9688226,
      size.height * 0.5814317, size.width * 0.9187136, size.height * 0.5326233);
  path_3.close();
  path.addPath(path_3, Offset.zero);

  Path path_4 = Path();
  path_4.moveTo(size.width * 0.6501858, size.height * 0.7249003);
  path_4.lineTo(size.width * 0.6501858, size.height * 0.9934280);
  path_4.lineTo(size.width * 0.7314723, size.height * 0.9934280);
  path_4.cubicTo(size.width * 0.8795396, size.height * 0.9934280, size.width,
      size.height * 0.8729676, size.width, size.height * 0.7249003);
  path_4.lineTo(size.width * 0.6501858, size.height * 0.7249003);
  path_4.close();

  path.addPath(path_4, Offset.zero);

  return path;
}


example/lib/main.dart
import 'dart:ui';

import 'package:example/new_svg.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:help_me/help_me.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  GlobalKey keyHello = GlobalKey();
  GlobalKey keyArrow = GlobalKey();
  GlobalKey keyHeadphone = GlobalKey();
  GlobalKey keyList = GlobalKey();
  GlobalKey keyButton = GlobalKey();

  @override
  Widget build(BuildContext context) {
    var appbar = AppBar(
      title: Text("Flutter HelpMe"),
    );
    return Scaffold(
      appBar: appbar,
      body: CustomScrollView(
        slivers: [
          SliverToBoxAdapter(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                SizedBox(
                  height: 24,
                ),
                Container(
                  key: keyHello,
                  padding: const EdgeInsets.all(8),
                  decoration: BoxDecoration(
                      borderRadius: BorderRadius.all(Radius.circular(8))),
                  child: const Text(
                    'Hello This is Example',
                    style: TextStyle(fontSize: 24, fontWeight: FontWeight.bold),
                  ),
                ),
                SizedBox(
                  height: 16,
                ),
                Container(
                  child: Text(
                    'A custom painter from svg',
                    textAlign: TextAlign.center,
                  ),
                ),
                SizedBox(
                  height: 32,
                ),
                Icon(Icons.arrow_downward_rounded, key: keyArrow),
                SizedBox(
                  height: 32,
                ),
              ],
            ),
          ),
          SliverToBoxAdapter(
            child: Column(
              children: [
                Container(
                    width: 100,
                    height: 100,
                    child: CustomPaint(
                      key: keyHeadphone,

                      // size: Size(50,
                      //     70), //You can Replace [WIDTH] with your desired width for Custom Paint and height will be calculated automatically
                      painter: RPSCustomPainter(),
                    )),
                SizedBox(
                  height: 24,
                ),
                Text(
                  "Test with a list!",
                  style: TextStyle(fontSize: 22),
                )
              ],
            ),
          ),
          SliverList(
              delegate: SliverChildBuilderDelegate((_, index) {
            return Card(
              key: index == 5 ? keyList : null,
              elevation: 10,
              color: Color.fromARGB(255, 58, 116, (index * 100) + 40),
              margin:
                  const EdgeInsets.only(right: 16, left: 16, bottom: 8, top: 8),
              child: Container(
                margin: const EdgeInsets.all(24),
                child: Text("Item : $index"),
              ),
            );
          }, childCount: 10)),
        ],
      ),
      floatingActionButton: FloatingActionButton.extended(
        key: keyButton,
        onPressed: () async {
          var currentBox =
              keyHeadphone.currentContext!.findRenderObject() as RenderBox;
          HelpMe.showHelp(
              gradient: LinearGradient(colors: [
                Colors.blue.withOpacity(0.2),
                Colors.purple.withOpacity(0.2),
              ]),
              items: [
                HelpMeItem(
                    key: keyButton,
                    shape: const HelpMeRectShape(
                        borderRadius: BorderRadius.all(Radius.circular(24))),
                    // shape:
                    //     HelperCustomShape(path: getPath(currentBox.size)),
                    guideWidget: Padding(
                      padding: const EdgeInsets.all(16.0),
                      child: Text(
                        "Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words, consectetur, from a Lorem Ipsum passage, and going through the cites of the word in classical literature, discovered the undoubtable source. Lorem Ipsum comes from sections 1.10.32 and 1.10.33 of  (The Extremes of Good and Evil) by Cicero, written in 45 BC. This book is a treatise on the theory of ethics, very popular during the Renaissance. The first line of Lorem Ipsum, comes from a line in section 1.10.32.The standard chunk of Lorem Ipsum used since the 1500s is reproduced below for those interested. Sections 1.10.32 and 1.10.33 from by Cicero are also reproduced in their exact original form, accompanied by English versions from the 1914 translation by H. Rackham.",
                        style: TextStyle(color: Colors.white),
                      ),
                    )),
                HelpMeItem(
                    key: keyHello,
                    shape: const HelpMeRectShape(
                        borderRadius: BorderRadius.all(Radius.circular(6))),
                    // shape:
                    //     HelperCustomShape(path: getPath(currentBox.size)),
                    guideWidget: Padding(
                      padding: const EdgeInsets.all(16.0),
                      child: Text(
                        "Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words, consectetur, from a Lorem Ipsum passage, and going through the cites of the word in classical literature, discovered the undoubtable source. Lorem Ipsum comes from sections 1.10.32 and 1.10.33 of  (The Extremes of Good and Evil) by Cicero, written in 45 BC. This book is a treatise on the theory of ethics, very popular during the Renaissance. The first line of Lorem Ipsum, comes from a line in section 1.10.32.The standard chunk of Lorem Ipsum used since the 1500s is reproduced below for those interested. Sections 1.10.32 and 1.10.33 from by Cicero are also reproduced in their exact original form, accompanied by English versions from the 1914 translation by H. Rackham.",
                        style: TextStyle(color: Colors.white),
                      ),
                    )),
                HelpMeItem(
                    key: keyArrow,
                    shape: HelpMeOvalShape(color: Colors.red, width: 8),
                    guideWidget: Padding(
                      padding: const EdgeInsets.all(16.0),
                      child: Column(
                        mainAxisSize: MainAxisSize.min,
                        children: [
                          Image.asset(
                            'assets/arrow.gif',
                            height: 60,
                          ),
                          Text(
                            "This is an arrow",
                            style: TextStyle(color: Colors.white),
                          ),
                        ],
                      ),
                    )),
                HelpMeItem(
                    key: keyHeadphone,
                    shape: HelpMeCustomShape(path: getPath2(currentBox.size)),
                    guideWidget: Padding(
                      padding: const EdgeInsets.all(16.0),
                      child: Text(
                        "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s",
                        style: TextStyle(color: Colors.white),
                      ),
                    )),
                HelpMeItem(
                    key: keyList,
                    shape: HelpMeRectShape(),
                    guideWidget: Padding(
                      padding: const EdgeInsets.all(16.0),
                      child: Text(
                        "Where can I get some?",
                        style: TextStyle(color: Colors.white),
                      ),
                    )),
              ],
              pageContext: context,
              nextWidget: Container(
                  padding: const EdgeInsets.only(
                      left: 24.0, right: 24, top: 8, bottom: 8),
                  margin: const EdgeInsets.all(8),
                  decoration: BoxDecoration(
                      color: Colors.purple,
                      borderRadius: BorderRadius.all(Radius.circular(8))),
                  child: Text(
                    "Next",
                    style: TextStyle(fontSize: 16, color: Colors.white),
                  )),
              skipWidget: Container(
                  padding: const EdgeInsets.only(
                      left: 24.0, right: 24, top: 8, bottom: 8),
                  margin: const EdgeInsets.all(8),
                  decoration: BoxDecoration(
                      color: Colors.purple,
                      borderRadius: BorderRadius.all(Radius.circular(8))),
                  child: Text(
                    "Skip",
                    style: TextStyle(fontSize: 16, color: Colors.white),
                  )));
        },
        tooltip: 'Increment',
        label: Text("Help!"),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
